ci: remove dependency on viz for some deep tests

mateiidavid opened the pull request:
Closes #7402 

Currently, we're trying to lower the time it takes for tests to run. As part of this effort, it would be great to remove the dependency our `deep` suite has on viz. Some tests can be moved out (particularly those that test viz functionality), while others can be changed to use `linkerd diagnostics` instead of viz counterparts to assert correct behavior w.r.t to connectivity.

Two viz dependent tests in our deep suite are `localhost_test` and `endpoints_test`. The former uses viz to get some metrics while the latter asserts endpoints exist for viz components.

For `localhost` through this change we would use regex and `linkerd diagnostics` to check the metrics. `endpoints_test` had its viz endpoints replaced with `nginx`.

Quick note for reviewers; for `endpoints_test` this change adds a bit of overhead, the assumption before in the test was that viz is already installed and running. Now, however, we have to install nginx and that adds a few seconds. I'd be in favour of removing nginx and only asserting that our control plane endpoints exist. For consistency, I wanted to check more than one namespace so I made the change, but I'm open to taking nginx out of the equation here.

<!--  Thanks for sending a pull request!

If you already have a well-structured git commit message, chances are GitHub
set the title and description of this PR to the git commit message subject and
body, respectively. If so, you may delete these instructions and submit your PR.

If this is your first time, please read our contributor guide:
https://github.com/linkerd/linkerd2/blob/main/CONTRIBUTING.md

The title and description of your Pull Request should match the git commit
subject and body, respectively. Git commit messages are structured as follows:

```
Subject

Problem

Solution

Validation

Fixes #[GitHub issue ID]

DCO Sign off
```

Example git commit message:

```
Introduce Pull Request Template

GitHub's community guidelines recommend a pull request template, the repo was
lacking one.

Introduce a `PULL_REQUEST_TEMPLATE.md` file.

Once merged, the
[Community profile checklist](https://github.com/linkerd/linkerd2/community)
should indicate the repo now provides a pull request template.

Fixes #3321

Signed-off-by: Jane Smith <jane.smith@example.com>
```

Note the git commit message subject becomes the pull request title.

For more details around git commits, see the section on Committing in our
contributor guide:
https://github.com/linkerd/linkerd2/blob/main/CONTRIBUTING.md#committing
-->

adleong reviewed (commented):
Big :+1: on this change.  Do you know if the CI failures are real or flakiness?

mateiidavid commented:
> Big 👍 on this change. Do you know if the CI failures are real or flakiness?

They were real this time around, I introduced an error in `endpoints_test`. We are using `endpoints_test` in two different suites: `deep` and `external-prometheus-deep`. The latter differs through an additional endpoint entry that it checks (for prometheus).

I changed the test a bit. If we run `external-prometheus` we do not check nginx endpoints. This is so we can avoid deploying nginx when running the external suite (nginx endpoints will be checked in deep anyway). We use different a different golden file when testing external-prometheus so it was quite easy to separate the two cases.

Also, small note for reviewers, please check my regex in `localhost_test` 🙏🏻

kleimkuhler reviewed (approved):
Awesome 👍

alpeb commented on test/integration/endpoints/endpoints_test.go:
```diff
@@ -20,27 +21,65 @@ func TestMain(m *testing.M) {
-		cmd = append(cmd, fmt.Sprintf("prometheus.%s.svc.cluster.local:9090", vizNs))
+		cmd = append(cmd, fmt.Sprintf("nginx.%s.svc.cluster.local:8080", "linkerd-endpoints-test"))
+		TestHelper.WithDataPlaneNamespace(ctx, "endpoints-test", map[string]string{}, t, func(t *testing.T, ns string) {
+			nginx, err := TestHelper.LinkerdRun("inject", "testdata/nginx.yaml")
+			if err != nil {
+				testutil.AnnotatedFatal(t, "unexpected error", err)
+			}
```
I think you can just add the inject annotation into the yaml directly to avoid this step...

mateiidavid replied:
Replaced with `TestHelper.Kubectl` so we avoid reading the file in (`TestHelper.KubectlApply` creates the deployment from stdin so we'd have to read the manifest in). Good suggestion 👍🏻!

alpeb commented on test/integration/endpoints/endpoints_test.go:
Isn't the external-prometheus case only relevant for the viz's suite, so not to be considered here?

mateiidavid replied:
Yup, you're right. I was thinking about this too. I think under our current plan to create better isolated test suites, we'd have to duplicate this test to have it present in both, `external` and `deep`. 

I don't think that's a really good idea, if we check the endpoints in `deep` then we can be pretty sure it's going to be the same regardless of what deployment in what namespace we have. From my perspective, having this endpoints test for external resources as well is a bit of a waste, we don't test anything new. I deleted the external prometheus part, I'm fine with testing just the control plane and nginx.

Wdyt?

alpeb replied:
Agreed, no benefit from having it in external resources too :+1:

alpeb reviewed (commented):
Looking good!

alpeb reviewed (approved):
Nice! :+1:
